src/main/java/com/aws/greengrass/builtin/services/mqttproxy/MqttProxyIPCAgent.java

```diff
@@ -101,16 +106,29 @@ public PublishToIoTCoreResponse handleRequest(PublishToIoTCoreRequest request) {
                 }
 
                 byte[] payload = validatePayload(request.getPayload(), serviceName);
-                QualityOfService qos = validateQoS(request.getQosAsString(), serviceName);
-                PublishRequest publishRequest = PublishRequest.builder().payload(payload).topic(topic).qos(qos).build();
-                CompletableFuture<Integer> future = mqttClient.publish(publishRequest);
+                com.aws.greengrass.mqttclient.v5.QOS qos = validateQoS(request.getQosAsString(), serviceName);
+                Publish publishRequest = Publish.builder()
+                        .payload(payload)
+                        .topic(topic)
+                        .qos(qos)
+                        .retain(Coerce.toBoolean(request.isRetain()))
+                        .correlationData(request.getCorrelationData())
+                        .responseTopic(request.getResponseTopic())
+                        .messageExpiryIntervalSeconds(request.getMessageExpiryIntervalSeconds())
+                        .userProperties(request.getUserProperties() == null ? null :
+                                request.getUserProperties().stream()
+                                        .map((u) -> new UserProperty(u.getKey(), u.getValue()))
+                                        .collect(Collectors.toList()))
+                        .payloadFormat(
+                                request.getPayloadFormat() == null || request.getPayloadFormat() == PayloadFormat.BYTES
+                                        ? Publish.PayloadFormatIndicator.BYTES : Publish.PayloadFormatIndicator.UTF8)
+                        .build();
 
-                // If the future is completed exceptionally then the MqttClient was unable to spool the request
                 try {
-                    future.getNow(0);
-                } catch (CompletionException e) {
+                    mqttClient.publish(publishRequest);
+                } catch (MqttRequestException | SpoolerStoreException | InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)